### PR TITLE
Introduced JWT token

### PR DESCRIPTION
**Pull request by ndedhia1210**
#### This PR consist of following changes
- Introduction of JWT Token
- From now on every service will require JWT to pass authorization 
- Login api is not deprecated for not impacting UI development
- Application will have to retrieve token from `auth` service
- Once token is retrieved, it needs to be used for making other service calls

#### Sample for Auth Service call
```
POST
http://localhost:3000/auth
Request: {
    "username": "mtripathi",
    "password": "mtripathi"
}
Response: "<JWT_Token>"
```

#### Sample for getting user information
```
GET
http://localhost:3000/getUser
Headers: {
    ...
    "authorization": "<JWT_Token>"
}
Response: { // Attribute in this response will be limited next PR
    "_id": "<UserId>",
    "name": "<Name>",
    "phone": "<Number>",
    "email": "<Email>",
    "address": "<Address>",
    "username": "<Username>",
    "password": "<Password>",
    "createdDate": "<CreatedDate>",
    "modifiedDate": "<ModifiedDate>"
}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,9 +11,11 @@ https://github.com/ndedhia1210/smallBusiness.git
 npm i
 ```
 
-- Set Mongo uri on local machine using (Please collect **MongoURI** this from MongoOwner)
+- Set Mongo uri on local machine using (Please collect **MongoURI, JWTTokenPrefix and JWTSecret** this from MongoOwner)
 ```
 export MONGO_CONNECTION_STRING="<MongoURI>"
+export JWT_TOKEN_PREFIX="<JWTTokenPrefix>"
+export JWT_SECRET="<JWTSecret>"
```
Comment (ndedhia1210) [inline README.md:18]: These environment values will be set on Server for security purpose.

File: app/auth/jwtTokensLib.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+const jwt = require('jsonwebtoken'); 
+const { sendError } = require('../util/error');
+
+// Following library is used for generating token
+// based on user detail
+exports.createToken = (user) => {
+    const accessToken = jwt.sign( 
+    { user: user._id }, 
+    process.env.JWT_SECRET // We can set TTL for this if needed
```
Comment (ndedhia1210) [inline app/auth/jwtTokensLib.js:9]: Let's discuss this in our next meeting.

File: app/auth/jwtTokensLib.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
+const jwt = require('jsonwebtoken'); 
+const { sendError } = require('../util/error');
+
+// Following library is used for generating token
+// based on user detail
+exports.createToken = (user) => {
+    const accessToken = jwt.sign( 
+    { user: user._id },
```
Comment (ndedhia1210) [inline app/auth/jwtTokensLib.js:8]: Planning to encrypt UserId in upcoming PR.

**Comment by swapnilkamat23** [comment]
Remove node_modules directory. Add it to .gitignore.

File: app/auth/jwtTokensLib.js (hunk 3 above)
Comment (swapnilkamat23) [reply app/auth/jwtTokensLib.js:8]: Sounds good

File: app/auth/jwtTokensLib.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    if (splitToken[0] !== process.env.JWT_TOKEN_PREFIX) 
+      return sendError(res, 401, "Invalid token!"); 
+    
+    const token = splitToken[1];
+    if (token === null) 
+      return res.status(400).json({ msg: "Missing token!" }); 
+
+    jwt.verify(token, process.env.JWT_SECRET, (err, user) => { 
+      if (err) return sendError(res, 401, "Invalid token!");  
+      req.user = user;
```
Comment (swapnilkamat23) [inline app/auth/jwtTokensLib.js:31]: Does this mean we have to send 'user' in request for every api call? Even for api that aren't user specific like getProductListByCategory, getProductDetails etc?

Comment (ndedhia1210) [reply app/auth/jwtTokensLib.js:31]: Not directly. 
like you will just have to pass Token for every service request. 

And in future I am planning to add user role along with user id and user role will be used for authorization. This is like one more layer of security. 

So basically 
- Token will help to authenticate request
- User role will help to Authorize user (user role=customer will not allowed to to call update product service) - Hence, if any intruder tries to invoke this service will be thrown error.

Comment (swapnilkamat23) [reply app/auth/jwtTokensLib.js:31]: Ok, I like that. Thanks for explaining!

**Review by swapnilkamat23 — APPROVED** [review]

**Pull request merged**